src/ctfeexpr.d

```diff
@@ -76,12 +76,12 @@ public:
     {
         ClassDeclaration cd = originalClass();
         uint fieldsSoFar = 0;
-        while (index - fieldsSoFar >= cd.fields.dim)
+        while ((value.elements.dim - fieldsSoFar - index) > cd.fields.dim)
         {
             fieldsSoFar += cd.fields.dim;
             cd = cd.baseClass;
         }
-        return cd.fields[index - fieldsSoFar];
+        return cd.fields[cd.fields.dim - (value.elements.dim - fieldsSoFar - index)];
     }
```

Comment:
I think we can use a logic similar to `NewExp::interpret` like:

```d
    VarDeclaration getFieldAt(uint index)
    {
        auto cd = originalClass();
        const size_t totalFieldCount = value.elements.dim;
        uint fieldsSoFar = totalFieldCount;
        assert(index < totalFieldCount);
        for (auto c = cd; c; c = c.baseClass)
        {
            fieldsSoFar -= c.fields.dim;
            if (fieldsSoFar <= index)
                return c.fields[index - fieldsSoFar];
        }
        assert(0);
    }
```

It's a little better than yours because:
1. The `totalFieldCount` is explicitly named.
2. If `index` is incorrectly larger than `totalFieldCount`, it would be properly asserted instead of access violation.